pypose/optim/solver.py

```diff
@@ -210,3 +213,112 @@ def forward(self, A: Tensor, b: Tensor) -> Tensor:
+    We recommend using only non-batched or batch size 1 input for this solver, as
+    the batched version was not appeared in the original scipy version. When handling
+    sparse matrices, the batched computation may introduce additional overhead.
+
+    Examples:
+        >>> import pypose.optim.solver as ppos
+        >>> A = torch.tensor([[0.1802967, 0.3151198, 0.4548111, 0.3860016, 0.2870615],
+                              [0.3151198, 1.4575327, 1.5533425, 1.0540756, 1.0795838],
+                              [0.4548111, 1.5533425, 2.3674474, 1.1222278, 1.2365348],
+                              [0.3860016, 1.0540756, 1.1222278, 1.3748058, 1.2223261],
+                              [0.2870615, 1.0795838, 1.2365348, 1.2223261, 1.2577004]])
+        >>> b = torch.tensor([[ 2.64306851],
+                              [-0.03593633],
+                              [ 0.73612658],
+                              [ 0.51501254],
+                              [-0.26689271]])
+        >>> A = A.to_sparse_csr() # optional, used to demonstrate the sparse matrix support
+        >>> solver = ppos.CG()
+        >>> x = solver(A, b)
+        tensor([[246.4098],
+                [ 22.6997],
+                [-56.9239],
+                [-161.7914],
+                [137.2683]])
+    '''
+    def __init__(self, maxiter=None, tol=1e-5):
+        super().__init__()
+        self.maxiter, self.tol = maxiter, tol
+
+    def forward(self, A: Tensor, b: Tensor, x: Optional[Tensor]=None,
+                M: Optional[Tensor]=None) -> Tensor:
+        '''
+        Args:
+            A (Tensor): the input tensor. It is assumed to be a symmetric
+                positive-definite matrix. Layout is allowed to be COO, CSR, BSR, or dense.
+            b (Tensor): the tensor on the right hand side. Layout could be sparse or dense
+                but is only allowed to be a type that is compatible with the layout of A.
+                In other words, `A @ b` operation must be supported by the layout of A.
+            x (Tensor, optional): the initial guess for the solution. Default: ``None``.
+            M (Tensor, optional): the preconditioner for A. Layout is allowed to be COO,
+                CSR, BSR, or dense. Default: ``None``.
+
+        Return:
+            Tensor: the solved tensor. Layout is the same as the layout of b.
+        '''
+        if A.ndim == b.ndim:
+            b = b.squeeze(-1)
+        else:
+            assert A.ndim == b.ndim + 1, \
+                'The number of dimensions of A and b must differ by 1'
+        if x is None:
+            x = torch.zeros_like(b)
+        bnrm2 = torch.linalg.norm(b, dim=-1)
+        if (bnrm2 == 0).all():
+            return b
+        atol = self.tol * bnrm2
+        n = b.shape[-1]
+
+        if self.maxiter is None:
+            maxiter = n*10
+        else:
+            maxiter = self.maxiter
+
+        matvec = partial(bmv, A)
```

Comment:
this makes codes longer, not necessary.

Reply:
matvec represents the A linear operator. Adding such variable would allow the reader to associate with scipy source code, as we do not have a linear operator type